### PR TITLE
ci: add autocorrect-lint workflow for zh-CN

### PR DESCRIPTION
**Pull request by huacnlee**
Continue for #3490, revert AutoCorrect again, improve CI for only check changed files.

### CODE REVIEW

**Comment by huacnlee** [comment]
CI example

image

https://github.com/huacnlee/translated-content/runs/5243996813?check_suite_focus=true

File: files/zh-cn/learn/index.md
Comment (irvin) [inline files/zh-cn/learn/index.md:10]: please rebase these un-need files

**Review by irvin — CHANGES_REQUESTED** [review]

**Comment by irvin** [comment]
@huacnlee if i understand https://github.com/mdn/translated-content/pull/3490#issuecomment-1042208948 correctly, we should limit the lint to run only on the changed file and only if they are under /zh-cn folder, did the patch include these restriction?

**Comment by huacnlee** [comment]
> @huacnlee if i understand [#3490 (comment)](https://github.com/mdn/translated-content/pull/3490#issuecomment-1042208948) correctly, we should limit the lint to run only on the changed file and only if they are under /zh-cn folder, did the patch include these restrictions?

I think zh-tw and jp, kr need to discussion, AutoCorrect can support them, but the MDN Core team consider to enable it.

File: files/zh-cn/learn/index.md
Comment (huacnlee) [reply files/zh-cn/learn/index.md:10]: This file change for show up the CI how to check the changed files and the CI execute result. 

If you have no question for the CI config, I can cancel this file change.

Comment (huacnlee) [reply files/zh-cn/learn/index.md:10]: CI result example:

https://github.com/mdn/translated-content/runs/5244123027?check_suite_focus=true

File: .github/workflows/autocorrect-lint.yml
Comment (huacnlee) [inline .github/workflows/autocorrect-lint.yml:15]: This means fetch recent 30 commits for get changed files.
In usually, a PR request commits times low that this number.

**Comment by huacnlee** [comment]
Hi @irvin @SphinxKnight, this Pull Request still waiting review to merge. 

I just updated AutoCorrect to newest version.

**Comment by yin1999** [comment]
I have a problem:

Does this workflow only lint the diff part, that means if I create a pr to modify a part of the document like this (#5600), will this workflow fail on it. I think it may confuse contributors (they have to modify the whole document to pass the lint test).

**Comment by irvin** [comment]
> I have a problem:
> 
> Does this workflow only lint the diff part, that means if I create a pr to modify a part of the document like this (#5600), will this workflow fail on it. I think it may confuse contributors (they have to modify the whole document to pass the lint test).

agree. if we could have it all fixed at once (like in #5726 ), then we can enable the lint for new contributions.

**Comment by SphinxKnight** [comment]
Sorry for the lag here, I'll have a look tomorrow

**Comment by SphinxKnight** [comment]
As for the GH Action file for this PR, it looks OK to me. We shall see after merging for undesired side effects, but I can't think of any now.

Regarding @yin1999's comment, I agree this might be a bit confusing for new contributors to get a red flag for something they didn't touch. I guess #5726 is one part of fixing this.
Alternatively, if reviewers know that a given change did not introduce a failure, they can decide to merge and open an issue for a fix on this doc afterwards.

**Review by SphinxKnight — APPROVED** [review]

**Comment by huacnlee** [comment]
I just updated CI to use the newest version of AutoCorrect.

And special the `.autocorrectignore` let it only enable for:

- files/zh-cn/
- docs/zh-cn/

image

@SphinxKnight @irvin @yin1999 I think, it's time to merge this PR now.

File: .autocorrectignore
Comment (huacnlee) [inline .autocorrectignore:5]: Disable this special file via:

https://github.com/mdn/translated-content/pull/8781#discussion_r982369678

Comment (yin1999) [reply .autocorrectignore:5]: https://github.com/mdn/translated-content/pull/8781#discussion_r985033499

Comment (caugner) [reply .autocorrectignore:5]: @yin1999 Is this a blocker? If so, what needs to happen before this PR can be merged?

Comment (huacnlee) [reply .autocorrectignore:5]: The AutoCorrect `textRules` new features can handle this.

https://github.com/huacnlee/autocorrect/releases/tag/v2.2.0

Can we add some special rules?

```
textRules:
  9.9亿: 0
  一二三,四五六.七八九: 0
  一二三,四五六,七八九,一二三,四五六,七八九: 0
```

Comment (yin1999) [reply .autocorrectignore:5]: Sounds better, what's your opinion @caugner

## background

In Chinese, we use `，` instead of `,` between short Chinese sentences, but the sentence metioned above is the output of JS API (See: <https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/BigInt/toLocaleString#using_locales>)

image

This tool will try to "correct" it.

Comment (caugner) [reply .autocorrectignore]: Sounds good to me, even though you will need to always bear this in mind when adding new code examples. Then again, you probably don't add examples with commas in their output everyday. 🙂

File: .github/workflows/autocorrect-lint.yml
Comment (yin1999) [inline .github/workflows/autocorrect-lint.yml:15]:
```suggestion
      - name: Get changed files
```

File: .github/workflows/autocorrect-lint.yml
Comment (yin1999) [inline .github/workflows/autocorrect-lint.yml:6]: Should we add a path filter here, just like this?

https://github.com/mdn/translated-content/blob/abc2a5d225cf19950e5778479b35a3f4efe3fcc9/.github/workflows/pr-test.yml#L13-L15

**Comment by huacnlee** [comment]
Recent check example:

image

**Comment by huacnlee** [comment]
@yin1999 All done.

image

And I will reset the last test commit, can we merge now?

File: .github/workflows/autocorrect-lint.yml
Comment (yin1999) [inline .github/workflows/autocorrect-lint.yml:34]: can we specific the version here. Just ref `main` may bring some broken changes.

**Review by yin1999 — CHANGES_REQUESTED** [review]

File: .github/workflows/autocorrect-lint.yml
Comment (huacnlee) [reply .github/workflows/autocorrect-lint.yml:34]: There still have issue in by AutoCorrect lint result. The space background highlight can't shown on GitHub Action result.

I am working for fix that, after that. I will lock it on a version.

Comment (huacnlee) [reply .github/workflows/autocorrect-lint.yml]: Version just locked.

Comment (huacnlee) [reply .github/workflows/autocorrect-lint.yml]: > There still have issue in by AutoCorrect lint result. The space background highlight can't shown on GitHub Action result.
> 
> I am working for fix that, after that. I will lock it on a version.

This but is GitHub Action CSS bug. 

https://github.com/huacnlee/github-actions-color-test

Comment (yin1999) [reply .github/workflows/autocorrect-lint.yml]: I'm not sure would a [problem matcher](https://github.com/orgs/community/discussions/25807) help with this. We have used it for [markdown lint](https://github.com/mdn/translated-content/blob/c070a9b09f1251ab6546514fd4b0e1b1ea026935/.github/workflows/markdownlint-problem-matcher.json).

File: .github/workflows/autocorrect-lint.yml
Comment (yin1999) [inline .github/workflows/autocorrect-lint.yml:26]: Let's remove this, see: mdn/content#22898

```suggestion
```

File: .github/workflows/autocorrect-lint.yml
Comment (yin1999) [inline .github/workflows/autocorrect-lint.yml:8]: I think we'd better add itself, and looks good to me now.

```suggestion
      - "files/zh-cn/**"
      - ".github/workflows/autocorrect-lint.yml"
```

**Review by yin1999 — APPROVED** [review]

**Comment by yin1999** [comment]
Hi @bsmth would you like to have a look on this?

**Comment by bsmth** [comment]
> Hi @bsmth would you like to have a look on this?

Hi @yin1999, sorry for the late reply. Taking a look 👀

File: .github/workflows/autocorrect-lint.yml
Comment (bsmth) [inline .github/workflows/autocorrect-lint.yml:24]: Maybe something to consider for future, it looks like we're [reusing this logic to check files](https://github.com/mdn/content/blob/221a88897598f0ad73f3e80c4a6211992264dcf5/.github/workflows/pr-check_markdownlint.yml#L22-L29) and it seems to be one of the more efficient ways of doing it after trial and error. We can consider adding it to the [shared workflows](https://github.com/mdn/workflows) repository where you can pass in a path for the `egrep` command. I can imagine us using it like:

```yml
  steps:
    - name: Get changed files
      uses: mdn/workflows/.github/workflows/list-changed-files.yml@main
      with:
        # don't run on forks
        target-repo: "mdn/translated-content"
        # only list these paths
        paths: "^files/zh-cn/"
```

**Review by bsmth — APPROVED** [review]
I don't have objections to adding this so I'm leaving a `+1`. Maybe @caugner would like to have a look? 👀

See my comment about sharing the duplication in the shared workflows repo for future improvement. 👍🏻

File: .github/workflows/autocorrect-lint.yml
Comment (yin1999) [reply .github/workflows/autocorrect-lint.yml:24]: Hi @bsmth. I did consider this before. And if I have understood [this](https://docs.github.com/en/actions/using-workflows/reusing-workflows). We could use this as a reusable job only. And in this job, we should set an output for using it in another job.

Or we should create a new action, so we could use this within a single job.

Comment (bsmth) [reply .github/workflows/autocorrect-lint.yml:24]: > In this job, we should set an output for using it in another job.

I think that's how the old action we were using did it, see https://github.com/tj-actions/changed-files/blob/1a3ea1d3240a68636af870fb623537087c4aaea2/action.yml#L111-L174

Comment (yin1999) [reply .github/workflows/autocorrect-lint.yml:24]: Yes, but it seems that GitHub only support reusable workflow, so we may have two jobs to deal with it:

```yml
jobs:
  get-changed-files:
    uses: mdn/workflows/.github/workflows/list-changed-files.yml@main
    with:
      target-repo: "mdn/translated-content"
      paths: "^files/zh-cn/"
    secrets:
      token: ${{ secrets.GITHUB_TOKEN }}

  some-job:
    runs-on: ubuntu-latest
    needs: get-changed-files
    if: ${{ needs.get-changed-files.outputs.files }}
    steps:
      # do sth.
```

Comment (bsmth) [reply .github/workflows/autocorrect-lint.yml:24]: Maybe this is good to record in a GitHub discussion

Comment (yin1999) [reply .github/workflows/autocorrect-lint.yml:24]: create a discussion as https://github.com/orgs/mdn/discussions/327

Comment (bsmth) [reply .github/workflows/autocorrect-lint.yml:24]: I'm going to resolve this as it's out of scope of the changes 👍🏻

**Comment by SphinxKnight** [comment]
+ @caugner

**Comment by bsmth** [comment]
pinging @LeoMcA for another set of eyes 👀

File: .autocorrectignore
Comment (caugner) [inline .autocorrectignore:3]:
```suggestion
files/
docs/
!files/zh-cn/

```

File: .autocorrectignore
Comment (caugner) [inline .autocorrectignore:1]: Can we add a comment here with a link to Autocorrect documentation?

File: .github/workflows/autocorrect-lint.yml
Comment (caugner) [inline .github/workflows/autocorrect-lint.yml:2]: Can we add a comment here explaining what autocorrect does (e.g. giving a very short example), link to https://huacnlee.github.io/autocorrect/, and mention that it is only enabled for `zh-CN`?

**Review by caugner — CHANGES_REQUESTED** [review]
LGTM, but let's add some more comments to provide context.

**Comment by yin1999** [comment]
Hi @huacnlee, do you have time to take a look :)

**Comment by huacnlee** [comment]
What can I need to do? Add the doc link?

File: .github/workflows/autocorrect-lint.yml
Comment (huacnlee) [reply .github/workflows/autocorrect-lint.yml]: Current only enabled for `zh-CN`, because that we have already fixed the zh-CN contents by a lot of pull request in before.

https://github.com/mdn/translated-content/pulls?q=is%3Apr+is%3Aclosed+author%3Ahuacnlee

But other translations (like `zh-TW`, `jp`) not done, there still have a lot of issues in the contents. We can start on `zh-CN` first and then gradually enable for other translations.

Comment (huacnlee) [reply .github/workflows/autocorrect-lint.yml]: I just write a description of this file.

File: .github/workflows/autocorrect-lint.yml
Comment (yin1999) [inline .github/workflows/autocorrect-lint.yml:5]:
```suggestion
# - incorrect: "欢迎阅读MDN文档."
```

File: .github/workflows/autocorrect-lint.yml
Comment (caugner) [inline .github/workflows/autocorrect-lint.yml:20]: Maybe add `.autocorrectrc` as well?

**Review by bsmth — APPROVED** [review]
Thanks a lot, I see all of @caugner's comments are addressed and I think we should merge.

Thanks everyone 👍🏻

**Comment by queengooborg** [comment]
I'm coming here after seeing CI failures in a few PRs I recently submitted to Prettier-ify the Simplified Chinese locale.  I think that this was merged a little prematurely.

While the CI test is great, there's no easy way to apply the changes it is requesting.  We have not added the package as a dependency, so we don't have easy access to the commands that allow us to locally test and fix reported issues.  **I should not have to rely on GitHub Actions to report issues in my commits; I should be able to run local, preconfigured commands to find all issues.**

I will submit a follow-up PR to add the local commands and integrate this into our local workflows soon.

**Comment by yin1999** [comment]
> I'm coming here after seeing CI failures in a few PRs I recently submitted to Prettier-ify the Simplified Chinese locale.  I think that this was merged a little prematurely.
> 
> While the CI test is great, there's no easy way to apply the changes it is requesting.  We have not added the package as a dependency, so we don't have easy access to the commands that allow us to locally test and fix reported issues.  **I should not have to rely on GitHub Actions to report issues in my commits; I should be able to run local, preconfigured commands to find all issues.**
> 
> I will submit a follow-up PR to add the local commands and integrate this into our local workflows, once I've worked out the issues.  In the meantime, I've disabled the workflow, and will re-enable it once we have set up a locally runnable command.

Yes. It's a better idea to add pre-commit hook to correct the format before committing. And we may need to fix the format of all zh-CN documents again.

**Pull request merged**
